include/spblas/vendor/rocsparse/multiply_spgemm.hpp

```diff
@@ -0,0 +1,326 @@
+    using output_type = std::remove_reference_t<decltype(c)>;
+    using value_type = typename matrix_type::scalar_type;
+
+    size_t buffer_size = 0;
+
+    auto alpha_optional = __detail::get_scaling_factor(a, b);
+    value_type alpha = alpha_optional.value_or(1);
+    auto beta_optional = __detail::get_scaling_factor(d);
+    value_type beta = beta_optional.value_or(1);
+    auto handle = this->handle_.get();
+    // Create sparse matrix A in CSR format
+    __rocsparse::throw_if_error(rocsparse_create_csr_descr(
+        &this->mat_a_, __backend::shape(a_base)[0], __backend::shape(a_base)[1],
+        a_base.values().size(), a_base.rowptr().data(), a_base.colind().data(),
+        a_base.values().data(),
+        to_rocsparse_indextype<typename matrix_type::offset_type>(),
+        to_rocsparse_indextype<typename matrix_type::index_type>(),
+        rocsparse_index_base_zero, to_rocsparse_datatype<value_type>()));
+    __rocsparse::throw_if_error(rocsparse_create_csr_descr(
+        &this->mat_b_, __backend::shape(b_base)[0], __backend::shape(b_base)[1],
+        b_base.values().size(), b_base.rowptr().data(), b_base.colind().data(),
+        b_base.values().data(),
+        to_rocsparse_indextype<typename input_type::offset_type>(),
+        to_rocsparse_indextype<typename input_type::index_type>(),
+        rocsparse_index_base_zero,
+        to_rocsparse_datatype<typename input_type::scalar_type>()));
+    __rocsparse::throw_if_error(rocsparse_create_csr_descr(
+        &this->mat_c_, __backend::shape(a_base)[0], __backend::shape(b_base)[1],
+        0, c.rowptr().data(), NULL, NULL,
+        to_rocsparse_indextype<typename output_type::offset_type>(),
+        to_rocsparse_indextype<typename output_type::index_type>(),
+        rocsparse_index_base_zero,
+        to_rocsparse_datatype<typename output_type::scalar_type>()));
+    __rocsparse::throw_if_error(rocsparse_create_csr_descr(
+        &this->mat_d_, __backend::shape(d_base)[0], __backend::shape(d_base)[1],
+        d_base.values().size(), d_base.rowptr().data(), d_base.colind().data(),
+        d_base.values().data(),
+        to_rocsparse_indextype<typename output_type::offset_type>(),
+        to_rocsparse_indextype<typename output_type::index_type>(),
+        rocsparse_index_base_zero,
+        to_rocsparse_datatype<typename output_type::scalar_type>()));
+    //--------------------------------------------------------------------------
+    // SpGEMM Computation
+    // ask buffer_size bytes for external memory
+    __rocsparse::throw_if_error(rocsparse_spgemm(
+        handle, rocsparse_operation_none, rocsparse_operation_none, &alpha,
+        this->mat_a_, this->mat_b_, &beta, this->mat_d_, this->mat_c_,
+        to_rocsparse_datatype<value_type>(), rocsparse_spgemm_alg_default,
+        rocsparse_spgemm_stage_buffer_size, &buffer_size, nullptr));
+    if (buffer_size > this->buffer_size_) {
+      this->alloc_.deallocate(workspace_, this->buffer_size_);
```

Comment:
let's have a reallocate method in the allocator?

Reply:
I think it is close to the array more than the allocator?
do you imagine something like `alloc_.reallocate(&workspace_, old_buffer_size_, new_buffer_size_)`?